README.md
Laravel Intl Formatter
=======================================

[![Author](http://img.shields.io/badge/author-@nyamsprod-blue.svg?style=flat-square)](https://twitter.com/nyamsprod)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE)
[![Build](https://github.com/bakame-php/laravel-intl-formatter/workflows/build/badge.svg)](https://github.com/bakame-php/laravel-intl-formatter/actions?query=workflow%3A%22build%22)
[![Latest Version](https://img.shields.io/github/release/bakame-php/laravel-intl-formatter.svg?style=flat-square)](https://github.com/bakame-php/laravel-intl-formatter/releases)
[![Total Downloads](https://img.shields.io/packagist/dt/bakame/laravel-intl-formatter.svg?style=flat-square)](https://packagist.org/packages/bakame/laravel-intl-formatter)
[![Sponsor development of this project](https://img.shields.io/badge/sponsor%20this%20package-%E2%9D%A4-ff69b4.svg?style=flat-square)](https://github.com/sponsors/nyamsprod)

This is a Laravel port of the [Twig Intl Extension](https://github.com/twigphp/intl-extra) package.

The package can be used in any Laravel based application to quickly handle internationalization 
by providing helper functions in Blade templates or Laravel codebase.

System Requirements
-------

- Laravel 8 and/or 9
- Symfony Intl component

Installation
------------

Use composer:

```
composer require bakame/laravel-intl-formatter
```

Configuration
------------

In order to edit the default configuration you need to publish the package configuration to your application config directory:

```bash
php artisan vendor:publish --provider="Bakame\Laravel\Intl" --tag=config
```

The configuration file will be published to `config/bakame-intl-formatter.php` in your application directory.

Please refer to the config file for an overview of the available options.

Documentation
------------

Once installed the package provides the following global helper functions:

### Country Name

Returns the country name given its two-letter/five-letter code;

```blade
country name: {{ country_name($country, $locale) }}
```

```php
echo view($templatePath, ['country' => 'FR', 'locale' => 'NL'])->render();
// country name: Frankrijk
```

### Currency Name

Returns the currency name given its three-letter code;

```blade
currency name: {{ currency_name($currency, $locale) }}
```

```php
echo view($templatePath, ['currency' => 'JPY', 'locale' => 'PT'])->render();
// currency name: Iene japonês
```

### Currency Symbol

Returns the currency symbol given its three-letter code;

```blade
currency symbol: {{ currency_symbol($currency, $locale) }}
```

```php
echo view($templatePath, ['currency' => 'JPY', 'locale' => 'PT'])->render();
// currency symbol: JP¥
```

### Language name

Returns the currency symbol given its three-letter code;

```blade
language name: {{ language_name($language, $locale) }}
```

```php
echo view($templatePath, ['language' => 'it', 'locale' => 'nl'])->render();
// language name: Italiaans
```

### Locale name

Returns the currency symbol given its three-letter code;

```blade
locale name: {{ locale_name($data, $locale) }}
```

```php
echo view($templatePath, ['data' => 'sw', 'locale' => 'nl'])->render();
// locale name: Swahili
```

### Timezone name

Returns the timezone name given its identifier;

```blade
timezone name: {{ locale_name($data, $locale) }}
```

```php
echo view($templatePath, ['timezone' => 'Asia/Tokyo', 'locale' => 'es'])->render();
// timezone name: hora de Japón (Tokio)
```

### Country Timezones

Returns the timezone identifiers of the given country code;

```blade
country timezones: {{ implde(", ", country_timezones($country)) }}
```

```php
$content = view($templatePath, ['country' => 'CD', 'locale' => 'es'])->render();
echo $content, PHP_EOL; // country timezones: Africa/Kinshasa, Africa/Lubumbashi
```

### Format Currency

Formats a number as a currency;

```blade
format currency: {{ format_currency($amount, $currency, $attrs, $locale) }}
```

```php
$templateData = [
    'amount' => 100.356, 
    'currency' => 'USD', 
    'locale' => 'ES', 
    'attrs' => [
        'fraction_digit' => 1,
        'rounding_mode' => 'floor',
    ]
];
echo view($templatePath, $templateData)->render();
// format currency: 100,3 US$
```

### Format Number

Formats a number;

```blade
format number: {{ format_number($number, $locale, $attrs) }}
```

```php
$templateData = [
    'number' => 100.356, 
    'locale' => 'nl', 
    'style' => 'spellout',
    'type' => 'double',
    'attrs' => [
        'fraction_digit' => 1,
        'rounding_mode' => 'floor',
    ]
];
echo view($templatePath, $templateData)->render();
// format number: honderd komma drie
```

### Format DateTime

Formats a date and time;

```blade
format datetime: {{ format_datetime($date, $locale, $timezone, $dateFormat, $timeFormat, $pattern, $calendar) }}
```

```php
$templateData = [
    'date' => 'yesterday', 
    'dateFormat' => 'full', 
    'timeFormat' => 'full', 
    'pattern' => '' ,
    'timezone' => 'Africa/Lubumbashi', 
    'calendar' => 'gregorian' ,
    'locale' => 'sw',
];
echo view($templatePath, $templateData)->render();
// format datetime: Alhamisi, 2 Juni 2022 00:00:00 Saa za Afrika ya Kati
```

### Format Date

Formats a the date portion of a datetime;

```blade
format date: {{ format_date($date, $locale, $timezone, $dateFormat, $pattern, $calendar) }}
```

```php
$templateData = [
    'date' => 'yesterday', 
    'dateFormat' => 'long', 
    'pattern' => '' ,
    'timezone' => 'Africa/Lubumbashi', 
    'calendar' => 'gregorian' ,
    'locale' => 'sw',
];
echo view($templatePath, $templateData)->render();
// format date: 2 Juni 2022
```

### Format Time

Formats the time portion of a datetime;

```blade
format time: {{ format_time($date, $locale, $timezone, $timeFormat, $pattern, $calendar) }}
```

```php
$templateData = [
    'date' => 'yesterday', 
    'dateFormat' => 'full', 
    'pattern' => '' ,
    'timezone' => 'Africa/Lubumbashi', 
    'calendar' => 'gregorian' ,
    'locale' => 'sw',
];
echo view($templatePath, $templateData)->render();
// format time: 00:00:00 Saa za Afrika ya Kati
```

Each function uses the same arguments in the same order as the Twig Extra package filters/functions.

### Locale specification 

If no `locale` is specified in function calls, the function will use the result of `Illuminate\Support\Facades\App::currentLocale()`
as the locale value to use.

### functions signature

In PHP8+, you can use named parameters to improve functions usages as they tend to have a lot of arguments:

**In PHP7.4**

```php 
<?php

echo format_datetime('2019-08-07 23:39:12', 'fr', null, medium', 'medium', '', 'gregorian', 'fr');
```

**In PHP8+**

```php 
<?php

echo format_datetime(date: '2019-08-07 23:39:12', locale: 'fr');
```

Contributing
-------

Contributions are welcome and will be fully credited. Please see [CONTRIBUTING](.github/CONTRIBUTING.md) and [CODE OF CONDUCT](.github/CODE_OF_CONDUCT.md) for details.

Testing
-------

The library:

- has a [PHPUnit](https://phpunit.de) test suite
- has a coding style compliance test suite using [PHP CS Fixer](https://cs.sensiolabs.org/).
- has a code analysis compliance test suite using [PHPStan](https://github.com/phpstan/phpstan).

To run the tests, run the following command from the project folder.

``` bash
$ composer test
```

Security
-------

If you discover any security related issues, please email nyamsprod@gmail.com instead of using the issue tracker.

Credits
-------

- [ignace nyamagana butera](https://github.com/nyamsprod)
- [All Contributors](https://github.com/bakame-php/laravel-intl-formatter/contributors)

Attribution
-------

The package helpers functions are heavily inspired by previous works done by [Fabien Potencier](https://github.com/fabpot) on [Twig Intl Extension](https://github.com/twigphp/intl-extra).

License
-------

The MIT License (MIT). Please see [License File](LICENSE) for more information.


CHANGELOG.md
# Changelog

All Notable changes to `bakame/laravel-intl-formatter` will be documented in this file

## [0.3.0] - 2022-06-14

### Added

- Support for `Money` package.
- Added `IntlFactory` facade.
- `NumberFactory` and `DateFactory` are now accessible as readonly property of the `Factory` class.

### Fixed

- **[BC Break]** Renamed `Factory::newInstance` in `Factory::newFormatter`
- **[BC Break]** Keys used for container registration are updated.

### Deprecated

- None

### Removed

**[BC Break]** `Factory::newInstance` replaced by `Factory::newFormatter`

## [0.2.0] - 2022-06-06

** Dependencies fixes

## [0.1.0] - 2022-06-04

**Initial release!**

[Next]: https://github.com/bakame-php/laravel-intl-formatter/compare/0.3.0...main
[Next]: https://github.com/bakame-php/laravel-intl-formatter/compare/0.2.0...0.3.0
[0.2.0]: https://github.com/bakame-php/laravel-intl-formatter/compare/0.1.0...0.2.0
[0.1.0]: https://github.com/bakame-php/laravel-intl-formatter/releases/tag/0.1.0


src/HelpersTest.php
<?php

declare(strict_types=1);

namespace Bakame\Intl\Laravel;

use Bakame\Intl\FailedFormatting;
use Illuminate\Contracts\Container\BindingResolutionException;
use Illuminate\Contracts\View\Factory as ViewFactory;
use Illuminate\Foundation\Application;
use Illuminate\Support\Facades\Artisan;
use Orchestra\Testbench\TestCase as LaravelTestCase;
use phpDocumentor\Reflection\Types\Void_;

final class HelpersTest extends LaravelTestCase
{
    public function setUp(): void
    {
        parent::setUp();

        Artisan::call('view:clear');
        Artisan::call('cache:clear');
    }

    /**
     * @param Application $app
     */
    protected function getEnvironmentSetUp($app): void
    {
        $app['config']->set('view.paths', [__DIR__.'/../test_files']);
    }

    /**
     * @param array<string, mixed> $withParameters
     *
     * @throws BindingResolutionException
     *
     * @return array<string>|string
     */
    public function renderView(string $viewName, array $withParameters = [])
    {
        return app(ViewFactory::class)->make($viewName, $withParameters)->render();
    }

    /**
     * @param Application $app
     *
     * @return array<int, class-string>
     */
    protected function getPackageProviders($app): array
    {
        return [
            Provider::class,
        ];
    }

    /** @test */
    public function it_can_format_a_country_name(): void
    {
        /** @var string $content */
        $content = $this->renderView('country_name');
        $expected = <<<EXPECTED
UNKNOWN

France
United States
États-Unis
Suisse
EXPECTED;
        self::assertStringContainsString($expected, $content);
    }

    /** @test */
    public function it_can_format_a_country_timezones(): void
    {
        /** @var string $content */
        $content = $this->renderView('country_timezones');
        $expected = <<<EXPECTED
0
Europe/Paris
America/Adak, America/Anchorage, America/Boise, America/Chicago, America/Denver, America/Detroit, America/Indiana/Knox, America/Indiana/Marengo, America/Indiana/Petersburg, America/Indiana/Tell_City, America/Indiana/Vevay, America/Indiana/Vincennes, America/Indiana/Winamac, America/Indianapolis, America/Juneau, America/Kentucky/Monticello, America/Los_Angeles, America/Louisville, America/Menominee, America/Metlakatla, America/New_York, America/Nome, America/North_Dakota/Beulah, America/North_Dakota/Center, America/North_Dakota/New_Salem, America/Phoenix, America/Sitka, America/Yakutat, Pacific/Honolulu
EXPECTED;
        self::assertStringContainsString($expected, $content);
    }

    /** @test */
    public function it_can_format_a_currency_name(): void
    {
        /** @var string $content */
        $content = $this->renderView('currency_name');
        $expected = <<<EXPECTED
UNKNOWN

Euro
Japanese Yen
euro
yen japonais
EXPECTED;
        self::assertStringContainsString($expected, $content);
    }

    /** @test */
    public function it_can_format_a_currency_symbol(): void
    {
        /** @var string $content */
        $content = $this->renderView('currency_symbol');
        $expected = <<<EXPECTED
UNKNOWN

€
¥
EXPECTED;
        self::assertStringContainsString($expected, $content);
    }

    /** @test */
    public function it_can_format_currency(): void
    {
        /** @var string $content */
        $content = $this->renderView('format_currency');
        $expected = <<<EXPECTED
€1,000,000.00
1.000.000,00 €
€1,000,000.00
€12.34
YEN 125,000.00
€12.34
EXPECTED;
        foreach (explode(PHP_EOL, $expected) as $expectedLine) {
            self::assertStringContainsString($expectedLine, $content);
        }
    }

    /** @test */
    public function it_can_format_date(): void
    {
        /** @var string $content */
        $content = $this->renderView('format_date');
        $expected = <<<EXPECTED
Aug 7, 2019, 11:39:12 PM
7 août 2019 à 23:39:12
23:39
07/08/2019
mercredi 7 août 2019 à 23:39:12 Temps universel coordonné
11 oclock PM, Coordinated Universal Time

Aug 7, 2019
7 août 2019
11:39:12 PM
EXPECTED;

        foreach (explode(PHP_EOL, $expected) as $expectedLine) {
            self::assertStringContainsString($expectedLine, $content);
        }
    }

    /** @test */
    public function it_can_format_number(): void
    {
        /** @var string $content */
        $content = $this->renderView('format_number');
        $expected = <<<EXPECTED
12.345
12,345
1,234%
twelve point three four five
1,234%
twelve point three four five
quatre-vingts virgule trois quatre cinq
huitante virgule trois quatre cinq
12 sec.
12.0%
13%
EXPECTED;
        self::assertStringContainsString($expected, $content);
    }

    /** @test */
    public function it_can_format_language_name(): void
    {
        /** @var string $content */
        $content = $this->renderView('language_name');
        $expected = <<<EXPECTED
UNKNOWN

German
French
allemand
français
français canadien
EXPECTED;
        self::assertStringContainsString($expected, $content);
    }

    /** @test */
    public function it_can_format_locale_name(): void
    {
        /** @var string $content */
        $content = $this->renderView('locale_name');
        $expected = <<<EXPECTED
UNKNOWN

German
French
allemand
français
français (Canada)
EXPECTED;
        self::assertStringContainsString($expected, $content);
    }

    /** @test */
    public function it_can_format_timezone_name(): void
    {
        /** @var string $content */
        $content = $this->renderView('timezone_name');
        $expected = <<<EXPECTED
UNKNOWN

Central European Time (Paris)
Pacific Time (Los Angeles)
heure du Pacifique nord-américain (Los Angeles)
EXPECTED;
        self::assertStringContainsString($expected, $content);
    }

    /** @test */
    public function it_throw_if_the_currency_is_missing(): void
    {
        $this->expectException(FailedFormatting::class);

        format_currency(300, null, 'nl_NL');
    }
}
